### PR TITLE
[New port] Rexo

### PR DESCRIPTION
**Pull request by SamuelMarks**
**Describe the pull request**

- #### What does your PR fix?

> [Rexo](https://github.com/christophercrouzet/rexo) is a neat single-file cross-platform unit testing framework for C/C++.
>
> It offers the same [xUnit](https://en.wikipedia.org/wiki/XUnit)-like structure than most other unit testing
> frameworks but aims at providing a _truly_ polished API.

- #### Which triplets are supported/not supported? Have you updated the [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt)?  
All

- #### Does your PR follow the [maintainer guide](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md)?  
Yes

- #### If you have added/updated a port: Have you run `./vcpkg x-add-version --all` and committed the result?  
Yes

**If you are still working on the PR, open it as a Draft: https://github.blog/2019-02-14-introducing-draft-pull-requests/**

### CODE REVIEW

**Comment by JonLiu1993** [comment]
@SamuelMarks ,Source provides cmakeLists, please use CMake to build, do not install header files directly.

**Comment by SamuelMarks** [comment]
@JonLiu1993 Even though it's a header-only single-file build which guarantees including the full header in its source tree?

**Comment by JackBoosY** [comment]
> @JonLiu1993 Even though it's a header-only single-file build which guarantees including the full header in its source tree?

Even if this port is just a header only library, the CMakeLists provided upstream also contain many other content, such as checks, export targets, tests, and installation examples.
So please use `vcpkg_cmake_configure` and `vcpkg_cmake_install` instead.
See https://github.com/christophercrouzet/rexo/blob/main/CMakeLists.txt#L244

File: versions/r-/rexo.json [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+    {
+      "git-tree": "ace24760ab40796719ace0912e5f8b90df60ec79",
+      "version-string": "0.2.0",
+      "port-version": 0
+    },
+    {
+      "git-tree": "e42b3790e03e0fecba47bd2eb9e68215b5ecc2ef",
+      "version-string": "2021-06-21",
+      "port-version": 0
+    }
```
Comment (JonLiu1993) [inline versions/r-/rexo.json:12]: Please delete this,we only need to add "version" once

**Review by JonLiu1993 — CHANGES_REQUESTED** [review]

File: versions/r-/rexo.json (hunk 1 above)
Comment (SamuelMarks) [reply versions/r-/rexo.json]: Updated. Also the issue I opened on the author's repo encouraged him to tag his releases. So this is tagged now.

File: ports/rexo/portfile.cmake [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+)
+
+vcpkg_configure_cmake(
+    SOURCE_PATH "${SOURCE_PATH}"
+    PREFER_NINJA
+    OPTIONS
+        -DREXO_BUILD_TESTS=OFF
+)
+
+file(COPY "${SOURCE_PATH}/include/rexo.h" DESTINATION "${CURRENT_PACKAGES_DIR}/include")
```
Comment (JackBoosY) [inline ports/rexo/portfile.cmake:16]:
```suggestion
vcpkg_cmake_install()
```

**Review by JackBoosY — CHANGES_REQUESTED** [review]

File: ports/rexo/CONTROL [hunk 3]
```diff
@@ -0,0 +1,4 @@
+Source: rexo
+Version: 0.2.0
+Description: Rexo is a neat single-file cross-platform unit testing framework for C/C++.
+Homepage: https://github.com/christophercrouzet/rexo
```
Comment (PhoebeHui) [inline ports/rexo/CONTROL:4]: Change this to vcpkg.json file, and let it depend on vcpkg-cmake port.

File: ports/rexo/portfile.cmake [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO "christophercrouzet/rexo"
+    REF "v0.2.0"
+    SHA512 "f22ca5be1b5d1a5159377609d2b666eaa69643281294202d2c37b8ab5c9d28acd5b67c2a3d5a0c4e6bae487b9ceff00a11dae1847e7a1e5f72193d9ad34b1457"
+    HEAD_REF "main"
+)
+
+vcpkg_configure_cmake(
+    SOURCE_PATH "${SOURCE_PATH}"
+    PREFER_NINJA
```
Comment (PhoebeHui) [inline ports/rexo/portfile.cmake:11]:
```suggestion
vcpkg_cmake_configure(
    SOURCE_PATH "${SOURCE_PATH}"
```

**Review by JonLiu1993 — CHANGES_REQUESTED** [review]
Change CONTROL file to vcpkg.json file:
```
{
  "name": "rexo",
  "version-semver": "0.2.0",
  "description": "Rexo is a neat single-file cross-platform unit testing framework for C/C++",
  "homepage": "https://github.com/christophercrouzet/rexo",
  "dependencies": [
    {
      "name": "vcpkg-cmake",
      "host": true
    }
  ]
}
```

File: ports/rexo/portfile.cmake [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+    OPTIONS
+        -DREXO_BUILD_TESTS=OFF
+)
+
+vcpkg_cmake_install()
+
```
Comment (JonLiu1993) [inline ports/rexo/portfile.cmake:17]:
```suggestion
vcpkg_fixup_cmake_targets(CONFIG_PATH lib/cmake/Rexo)

file(REMOVE_RECURSE "${CURRENT_PACKAGES_DIR}/debug" "${CURRENT_PACKAGES_DIR}/lib")
```

File: ports/rexo/portfile.cmake [hunk 6]
```diff
@@ -0,0 +1,18 @@
+vcpkg_from_github(
+    OUT_SOURCE_PATH SOURCE_PATH
+    REPO "christophercrouzet/rexo"
+    REF "v0.2.0"
+    SHA512 "f22ca5be1b5d1a5159377609d2b666eaa69643281294202d2c37b8ab5c9d28acd5b67c2a3d5a0c4e6bae487b9ceff00a11dae1847e7a1e5f72193d9ad34b1457"
+    HEAD_REF "main"
```
Comment (JonLiu1993) [inline ports/rexo/portfile.cmake:6]:
```suggestion
    HEAD_REF "main"
    PATCHES
        fix-include-path.patch
```

**Review by JonLiu1993 — CHANGES_REQUESTED** [review]

**Comment by JonLiu1993** [comment]
fix-include-path.patch:
```
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 2dc73f2..fb4d990 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -234,7 +234,7 @@ endif()
 # ------------------------------------------------------------------------------
 
 install(
-    DIRECTORY include
+    DIRECTORY include/
     DESTINATION ${CMAKE_INSTALL_INCLUDEDIR})
 
 # ------------------------------------------------------------------------------

```

**Comment by JonLiu1993** [comment]
@SamuelMarks，vcpkg_cmake_configure and vcpkg_cmake_install provided by vcpkg-cmake port.so you should Change CONTROL file to vcpkg.json file:
````
{
  "name": "rexo",
  "version-semver": "0.2.0",
  "description": "Rexo is a neat single-file cross-platform unit testing framework for C/C++",
  "homepage": "https://github.com/christophercrouzet/rexo",
  "dependencies": [
    {
      "name": "vcpkg-cmake",
      "host": true
    }
  ]
}
````

**Comment by JonLiu1993** [comment]
@SamuelMarks ,Could ypu please take a look:
https://dev.azure.com/vcpkg/public/_build/results?buildId=56366&view=artifacts&pathAsName=false&type=publishedArtifacts

**Comment by PhoebeHui** [comment]
Ping @SamuelMarks for response, Is work still being done for this PR?

File: ports/rexo/CONTROL (hunk 3 above)
Comment (JonLiu1993) [reply ports/rexo/CONTROL]: @SamuelMarks ,Could you please adopt the above comments?

**Comment by JackBoosY** [comment]
Get failure logs [here](https://dev.azure.com/vcpkg/public/_build/results?buildId=57250&view=artifacts&pathAsName=false&type=publishedArtifacts).

**Comment by SamuelMarks** [comment]
Oh I forgot to update the SHA hash

File: ports/rexo/CONTROL [hunk 7]
```diff
@@ -0,0 +1,4 @@
+Source: rexo
+Version: 0.2.2
+Description: Rexo is a neat single-file cross-platform unit testing framework for C/C++.
+Homepage: https://github.com/christophercrouzet/rexo
```
Comment (JackBoosY) [inline ports/rexo/CONTROL:4]: Can you please use manifest mode (convert CONTROL to vcpkg.json)? See [documentation](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/manifest-files.md#manifest-files---vcpkgjson).

**Review by JackBoosY — CHANGES_REQUESTED** [review]

File: ports/rexo/CONTROL (hunk 7 above)
Comment (SamuelMarks) [reply ports/rexo/CONTROL:4]: Why? - And if it's the new format then why not a simple script to convert all the ports from `CONTROL` to `vcpkg.json`? - On that note, if I wrote such a script would you accept a PR which modified this for all `CONTROL` using ports?

Comment (JackBoosY) [reply ports/rexo/CONTROL:4]: Since this is a new port, we prefer to use manifest file. And `CONTROL` will be deprecated in the future.

Comment (PhoebeHui) [reply ports/rexo/CONTROL]: Another reason is [vcpkg_configure_cmake](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/vcpkg_configure_cmake.md), [vcpkg_install_cmake](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/vcpkg_install_cmake.md) and [vcpkg_fixup_cmake_targets](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/vcpkg_fixup_cmake_targets.md) has been deprecated, and the new functions requires to depend on port vcpkg-cmake and vcpkg-cmake-config, which used in vcpkg.json file.

File: ports/rexo/vcpkg.json [hunk 8]
```diff
@@ -0,0 +1,6 @@
+{
+  "name": "rexo",
+  "version-semver": "0.2.2",
+  "description": "Rexo is a neat single-file cross-platform unit testing framework for C/C++",
+  "homepage": "https://github.com/christophercrouzet/rexo"
```
Comment (JonLiu1993) [inline ports/rexo/vcpkg.json:5]:
```suggestion
  "homepage": "https://github.com/christophercrouzet/rexo"
  "dependencies": [
    {
      "name": "vcpkg-cmake",
      "host": true
    }
  ]
```

**Review by JonLiu1993 — CHANGES_REQUESTED** [review]

File: ports/rexo/vcpkg.json (hunk 8 above)
Comment (SamuelMarks) [reply ports/rexo/vcpkg.json:5]: Does it depend on that though? - Isn't that the other syntax which I reverted, `vcpkg_configure_cmake` vs `vcpkg_cmake_configure`?

Comment (JonLiu1993) [reply ports/rexo/vcpkg.json:5]: > Does it depend on that though? - Isn't that the other syntax which I reverted, `vcpkg_configure_cmake` vs `vcpkg_cmake_configure`?

Sorry, I made a mistake, if you use vcpkg_configure_cmake, you don't need to add this dependency

Comment (JonLiu1993) [reply ports/rexo/vcpkg.json:5]: Could you please replace old functions with new cmake functions?

Comment (PhoebeHui) [reply ports/rexo/vcpkg.json:5]: Yes, the new functions 'vcpkg_cmake_configure' provided via port vcpkg-cmake. So when you use vcpkg_cmake_configure, you need add vcpkg-cmake as a dependency. eg: https://github.com/microsoft/vcpkg/blob/master/ports/recast/vcpkg.json

See https://github.com/microsoft/vcpkg/pull/18895#discussion_r672899520

**Comment by JonLiu1993** [comment]
@SamuelMarks ,Could you modify the pr according to the review comments above?

**Comment by SamuelMarks** [comment]
@JonLiu1993 I asked a question on August 5th but didn't get a response…

**Comment by JackBoosY** [comment]
`vcpkg_*_cmake` will be deprecated, please use `vcpkg_cmake_*` and add `vcpkg-cmake` as a host dependency.

**Comment by JonLiu1993** [comment]
> @JonLiu1993 I asked a question on August 5th but didn't get a response…

 The new functions 'vcpkg_cmake_configure' provided via port vcpkg-cmake. So when you use vcpkg_cmake_configure, you need add vcpkg-cmake as a dependency

**Comment by JonLiu1993** [comment]
@SamuelMarks ，After finish your suggesstion changes,please run command "./vcpkg x-add-version rexo --overwrite-version" then commit the changes again

**Review by JonLiu1993 — APPROVED** [review]

**Comment by JonLiu1993** [comment]
@SamuelMarks ,Thanks for your contribution.

**Review by JackBoosY — APPROVED** [review]

**Comment by BillyONeal** [comment]
Thanks for the new port :D

**Pull request merged**